### PR TITLE
Fixing addLater component lifecycle

### PR DESCRIPTION
**Pull request by erickzanardo**
When using `BaseGame#addLater`, `onMount` on Component was not called, causing problematic behaviours.

### CODE REVIEW

File: lib/game/game.dart [hunk 1]
```diff
@@ -229,7 +229,7 @@ abstract class BaseGame extends Game with TapDetector {
   /// You can override it further to add more custom behaviour.
   @override
   void update(double t) {
-    components.addAll(_addLater);
+    _addLater.forEach(add);
```
Comment (luanpotter) [inline lib/game/game.dart:232]: this doesn't look right to me. preAdd is called immediately when the component is added to the add later list. you are calling it twice by calling add here. I separated all handling to `preAdd` so we didn't have this problem.

Comment (erickzanardo) [reply lib/game/game.dart:232]: This makes sense, I don't know how I missed the `predAdd` method...

**Comment by erickzanardo** [comment]
Updated @luanpotter

**Comment by spydon** [comment]
Wouldn't it be cleaner to iterate over the _addLater list in update and send in the add function?
So that we don't have to have to separate code paths.
Something like:
`_addLater.foreach(add)`

**Comment by erickzanardo** [comment]
> Wouldn't it be cleaner to iterate over the _addLater list in update and send in the add function?
> So that we don't have to have to separate code paths.
> Something like:
> `_addLater.foreach(add)`
> 
> EDIT: Oh, that's what it used to do. Can't you just move out preAdd out of the addLater() logic to avoid the double call?

@luanpotter I kind of like this idea, what do you think?

**Comment by spydon** [comment]
> LGTM

What do you think about removing the preAdd call from the addLater() logic and just do _addLater.foreach(add) in update? So that we can have two slightly less different different code paths for the adding logic.

**Comment by luanpotter** [comment]
@spydon that is a a good idea, IIRC I started doing that but I changed to call preAdd prematurely because one of my components was expecting things to be initialized right away, even before being actually added to the game. I don't recall the details rn but as long as our components keep working even with the delayed call to preAdd, I think it's ok to change.

**Comment by luanpotter** [comment]
@spydon @erickzanardo let's merge this as is and we can reconsider whether to join preAdd to add or not on a followup. I actually think it might be valuable to be able to override each one individually as you might want to hook on preAdd (maybe should be called prepare or something) and also independently be able to replace the component list with another impl of ordered set by override add.

**Pull request merged**